File: PitBoard(Android)/app/src/main/java/com/pitboard/app/util/EventWeekendGrouper.kt
package com.pitboard.app.util

import com.pitboard.app.data.EventEntity
import java.time.DayOfWeek
import java.time.Instant
import java.time.LocalDate
import java.time.ZoneId
import java.time.temporal.TemporalAdjusters

data class EventWeekendGroups(
    val weekendLabel: String,
    val weekendEvents: List<EventEntity>,
    val laterEvents: List<EventEntity>,
)

object EventWeekendGrouper {

    /**
     * Separa eventos en el bloque del fin de semana más cercano que contenga eventos
     * (viernes–domingo) y el resto de eventos futuros.
     */
    fun split(events: List<EventEntity>, zone: ZoneId = ZoneId.systemDefault()): EventWeekendGroups {
        if (events.isEmpty()) {
            return EventWeekendGroups(weekendLabel = "", weekendEvents = emptyList(), laterEvents = emptyList())
        }

        // Buscamos el primer evento para determinar cuál es el "fin de semana más próximo" con actividad
        val firstDate = Instant.ofEpochMilli(events.first().startTimeUtc).atZone(zone).toLocalDate()
        
        // Definimos el margen del fin de semana (Viernes a Domingo) que envuelve a ese primer evento
        val friday = firstDate.with(TemporalAdjusters.previousOrSame(DayOfWeek.FRIDAY))
        val sunday = firstDate.with(TemporalAdjusters.nextOrSame(DayOfWeek.SUNDAY))
        
        val weekendStartMs = friday.atStartOfDay(zone).toInstant().toEpochMilli()
        val weekendEndMs = sunday.plusDays(1).atStartOfDay(zone).toInstant().toEpochMilli() - 1

        val today = LocalDate.now(zone)
        val label = when {
            firstDate == today -> "Hoy"
            today in friday..sunday -> "Este fin de semana"
            friday == today.with(TemporalAdjusters.next(DayOfWeek.FRIDAY)) -> "Próximo fin de semana"
            else -> "Próxima cita"
        }

        val weekend = events.filter { it.startTimeUtc in weekendStartMs..weekendEndMs }
        val later = events.filter { it.startTimeUtc > weekendEndMs }

        return EventWeekendGroups(
            weekendLabel = label,
            weekendEvents = weekend,
            laterEvents = later
        )
    }
}

File: PitBoard(Iphone)/PitBoardKit/Sources/PitBoardKit/Support/EventWeekendGrouper.swift
import Foundation

public struct EventWeekendGroups: Sendable {
    public var weekendLabel: String
    /// Misma etiqueta que `weekendLabel` pero como clave de `Strings` (ej.
    /// "events_weekend_today") en vez de texto fijo en español — `weekendLabel` se deja
    /// igual a propósito (los tests existentes y cualquier otro consumidor lo siguen
    /// leyendo tal cual) y esta clave nueva es solo para quien quiera traducirla (ver
    /// EventsScreen.swift). Vacía cuando `weekendLabel` también lo está (sin eventos).
    public var weekendLabelKey: String
    public var weekendEvents: [EventModel]
    public var laterEvents: [EventModel]
}

/// Separa eventos en el bloque del fin de semana más cercano que contenga eventos
/// (viernes–domingo) y el resto de eventos futuros — equivalente exacto de
/// `EventWeekendGrouper.kt`.
public enum EventWeekendGrouper {
    public static func split(_ events: [EventModel], zone: TimeZone = .current, now: Date = Date()) -> EventWeekendGroups {
        guard let first = events.first else {
            return EventWeekendGroups(weekendLabel: "", weekendLabelKey: "", weekendEvents: [], laterEvents: [])
        }

        var calendar = Calendar(identifier: .gregorian)
        calendar.timeZone = zone

        let firstDate = calendar.startOfDay(for: first.startTimeUtc)
        // Calendar Gregoriano: weekday 1 = domingo ... 6 = viernes ... 7 = sábado.
        let firstWeekday = calendar.component(.weekday, from: firstDate)
        let daysBackToFriday = (firstWeekday - 6 + 7) % 7
        let daysForwardToSunday = (1 - firstWeekday + 7) % 7

        guard
            let friday = calendar.date(byAdding: .day, value: -daysBackToFriday, to: firstDate),
            let sunday = calendar.date(byAdding: .day, value: daysForwardToSunday, to: firstDate),
            let sundayEnd = calendar.date(byAdding: .day, value: 1, to: sunday).map({ $0.addingTimeInterval(-1) })
        else {
            return EventWeekendGroups(weekendLabel: "", weekendLabelKey: "", weekendEvents: [], laterEvents: [])
        }

        let today = calendar.startOfDay(for: now)
        let label: String
        let labelKey: String
        if firstDate == today {
            label = "Hoy"
            labelKey = "events_weekend_today"
        } else if today >= friday && today <= sunday {
            label = "Este fin de semana"
            labelKey = "events_weekend_this"
        } else if friday == nextFriday(strictlyAfter: today, calendar: calendar) {
            label = "Próximo fin de semana"
            labelKey = "events_weekend_next"
        } else {
            label = "Próxima cita"
            labelKey = "events_weekend_upcoming"
        }

        let weekend = events.filter { $0.startTimeUtc >= friday && $0.startTimeUtc <= sundayEnd }
        let later = events.filter { $0.startTimeUtc > sundayEnd }

        return EventWeekendGroups(weekendLabel: label, weekendLabelKey: labelKey, weekendEvents: weekend, laterEvents: later)
    }

    /// Equivalente de `TemporalAdjusters.next(DayOfWeek.FRIDAY)`: el próximo viernes
    /// ESTRICTAMENTE después de `date` — si `date` ya es viernes, salta a la semana
    /// siguiente (a diferencia de `nextOrSame`).
    private static func nextFriday(strictlyAfter date: Date, calendar: Calendar) -> Date {
        let weekday = calendar.component(.weekday, from: date)
        let daysForward = (6 - weekday + 7) % 7
        let offset = daysForward == 0 ? 7 : daysForward
        return calendar.date(byAdding: .day, value: offset, to: date) ?? date
    }
}
